site/src/routes/pins-add.js

```diff
@@ -3,43 +3,114 @@ import { verifyToken } from '../utils/utils.js'
  * @returns {Promise<JSONResponse>}
  */
-export async function pinsAdd (event) {
+export async function pinsAdd(event) {
   const result = await verifyToken(event)
   if (!result.ok) {
     return HTTPError.respond(result.error)
   }
   const { user, tokenName } = result
   const pinData = await event.request.json()
-  pinData.meta = pinData.meta || {}
-  pinData.meta.sub = user.sub
 
-  const res = await PinataPSA.pinsAdd(pinData)
-  if (!res.ok) {
-    return PinataPSA.parseErrorResponse(res.error)
+  if (typeof pinData.cid !== 'string') {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'invalid CID' } },
+      { status: 400 }
+    )
+  }
+  if (pinData.name && typeof pinData.name !== 'string') {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'invalid name' } },
+      { status: 400 }
+    )
+  }
+
+  const name = pinData.name
+
+  let meta
+  if (pinData.meta) {
+    if (typeof pinData.meta !== 'object' || Array.isArray(pinData.meta)) {
+      return new JSONResponse(
+        { error: { reason: 'INVALID_PIN_DATA', details: 'invalid metadata' } },
+        { status: 400 }
+      )
+    }
+    meta = Object.fromEntries(
+      Object.entries(pinData.meta).filter(([, v]) => typeof v === 'string')
+    )
   }
 
-  const pinStatus = res.value
+  /** @type import('@nftstorage/ipfs-cluster').PinResponse | undefined */
+  let pin
+  /** @type import('../pinata-psa').Status */
+  let status = 'pinning'
+  let size = 0
+  try {
+    pin = await cluster.allocation(pinData.cid)
+  } catch (err) {
+    // if 404 we can continue, we have not pinned this CID
+    if (!err.response || err.response.status !== 404) {
+      throw err
+    }
+  }
 
+  if (pin) {
+    // if we cached the size then use it
+    if (pin.metadata && pin.metadata.size) {
+      size = parseInt(pin.metadata.size)
+    }
+    // we have this pin already allocated in our cluster, get the status...
+    const { peerMap } = await cluster.status(pin.cid)
+    status = cluster.toPSAStatus(Object.values(peerMap))
+    // if this failed to pin, try again
+    if (status === 'failed') {
+      await cluster.recover(pinData.cid)
+    }
+  } else {
+    pin = await cluster.pin(pinData.cid)
+  }
+
+  event.waitUntil(PinataPSA.pinsAdd(pinData))
+
+  const created = new Date()
+  /** @type import('../bindings').NFT */
   const nft = {
-    cid: pinStatus.pin.cid,
-    size: 0,
-    created: new Date().toISOString(),
+    cid: pin.cid,
+    size,
+    created: created.toISOString(),
     type: 'remote',
     scope: tokenName,
     files: [],
     pin: {
-      cid: pinStatus.pin.cid,
-      size: 0,
-      status: pinStatus.status,
-      created: pinStatus.created
-    }
+      cid: pin.cid,
+      name,
+      meta,
+      size,
+      status,
+      created: created.toISOString(),
+    },
+  }
+  const metadata = {
+    name,
+    meta,
+    pinStatus: status,
+    size,
+    created: created.toISOString(),
   }
```

Comment:
we duplicate a bunch of stuff here, we should only keep the status and created in the metadata because we need that to filter in the list endpoint and deduplicate the rest inside the kv value

Reply:
i guess psa filters by meta too..

Reply:
Yep, and then we also need `size` as it is used by the metrics endpoint in order to count the total size of all NFTs.